BlueprintUI/Sources/Element/ElementState.swift

```diff
@@ -542,6 +632,68 @@ extension Dictionary {
     }
 }
 
+//
+// MARK: ElementStateTreeDelegate
+//
+
+/// A delegate object that is called during layout to inspect the actions performed
+/// during the layout and measurement process.
+protocol ElementStateTreeDelegate: AnyObject {
+
+    /// Root `ElementState`
+
+    func tree(_ tree: ElementStateTree, didSetupRootState state: ElementState)
+    func tree(_ tree: ElementStateTree, didUpdateRootState state: ElementState)
+    func tree(_ tree: ElementStateTree, didTeardownRootState state: ElementState)
+    func tree(_ tree: ElementStateTree, didReplaceRootState state: ElementState, with new: ElementState)
+
+    /// Creating / Updating `ElementState`
+
+    func treeDidCreateState(_ state: ElementState)
+    func treeDidUpdateState(_ state: ElementState)
+    func treeDidRemoveState(_ state: ElementState)
+
+    func treeDidFetchElementContent(for state: ElementState)
+
+    /// Measuring & Laying Out
+
+    func treeDidReturnCachedMeasurement(
+        _ measurement: ElementState.CachedMeasurement,
+        for state: ElementState
+    )
+
+    func treeDidPerformMeasurement(
+        _ measurement: ElementState.CachedMeasurement,
+        for state: ElementState
+    )
+
+    func treeDidReturnCachedLayout(
+        _ layout: ElementState.CachedLayout,
+        for state: ElementState
+    )
+
+    func treeDidPerformLayout(
+        _ layout: [LayoutResultNode],
+        for state: ElementState
+    )
+
+    func treeDidPerformCachedLayout(
+        _ layout: ElementState.CachedLayout,
+        for state: ElementState
+    )
+}
+
+
+extension Optional where Wrapped == ElementStateTreeDelegate {
+    func perform(_ block: (Wrapped) -> Void) {
```

Comment:
Could we name this function in a way where it's clear that it's only performed in `DEBUG` mode? `ifDebugPerform`?

Reply:
How about just `ifDebug`?

Reply:
This is done in a future branch